streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java

```diff
@@ -609,17 +842,18 @@ public void testOrdering() {
             inputTopic1.pipeInput(1, "A1", 100L);
             processor.checkAndClearProcessResult();
 
-            // push one item to the other window that has a join; this should produce non-joined records with a closed window first, then
-            // the joined records
+            // push one item to the other window that has a join; 
+            // this should produce the joined record first;
+            // then non-joined record with a closed window
             // by the time they were produced before
             // w1 = { 0:A0 (ts: 0), 1:A1 (ts: 100) }
             // w2 = { }
             // --> w1 = { 0:A0 (ts: 0), 1:A1 (ts: 100) }
             // --> w2 = { 1:a1 (ts: 110) }
             inputTopic2.pipeInput(1, "a1", 110L);
             processor.checkAndClearProcessResult(
-                new KeyValueTimestamp<>(0, "A0+null", 0L),
-                new KeyValueTimestamp<>(1, "A1+a1", 110L)
+                new KeyValueTimestamp<>(1, "A1+a1", 110L),
+                new KeyValueTimestamp<>(0, "A0+null", 0L)
```

Comment:
It seems your change to push out "left/outer" join result at the end of `process()` instead of at the beginning leads to this revers output, what implies we now emit data out-of-order. -- Hence, it seems better to keep the code as-is?

Reply:
Can you point me to the ordering-rules in case of a  left- outer- inner- join, then I can have a look.

Reply:
This is because previously we look at outer store and then join. But this change make it we join first and then look at outer store. The ts in outer store and other store is hard to reason. If we change the ts of 0 to be 100 and ts of 1 to be 50, the original test would still produce 0 first which has larger ts... So unless we compare the ts of join and outer at the same time when we output, we can guarantee the order of ts when output.

Reply:
Agreed

Reply:
What we have seen is that you could emit both a null-joined record and a joined-record if you first check the outerjoinstore  (via emitNonJoinedOuterRecords()) and only after that, a record from the otherWindowStore. 
This behaviour is not what you want.
Therefore you should first find out if there is going to be a joined record and if there is, you have to nullify this in the relevant outerjoinstore.

Reply:
I agree that we should not produce a null-joined record and joined-record. Can you elaborate on the case when (and why) this could exactly happen? It sounds like a bug in `emitNonJoinedOuterRecords()` to me.

When we check the `outerStore` we should only emit null-joined-records that are "expired", ie, which belong to already closed windows, and thus, the order in which we check should not matter. It sounds like as if we might incorrectly emit null-joined records of window that are not closed yet (would be a bug) or for windows which are already closed but we emit an incorrect join-record.

Given that we identified that we indeed have a bug with regard to "late records" and that we don't respect the grace-period correctly, I would assume it's the same root cause.

Thus, I am now wondering if we should merge this PR as-is (and fix the order back in a follow PR @florin-akermann is doing), or revert this change in this PR right away, or first try to merge Florin's PR and re-evaluate this change afterwards?

Thoughts?

Reply:
When we change the order in which we check back to as it was, only the ordering tests will fail. So there seems to be no bug (anymore).
I think that in a previous version of this PR (before checking both the left and right in the outerjoin) we saw that it did matter and a left-join test failed. But this has been solved now with the getOuterJoinLookBackTimeMs().
I think option 2 and option 3 will do fine.

Reply:
Seems this PR is basically ready for merging, to it might be faster to go with option (2), and revert changing the order in this PR and we can merge it.